projects/hathor-wallet-mobile/0003-token-swap.md

```diff
@@ -0,0 +1,98 @@
+
+The same should apply when the user chooses the output token value, he should deposit more than the expected value to account for slippage and any outstanding amount will be left as balance on the contract.
+
+This means that the accepted slippage will be defined by the user's wallet and the swap will automatically fail if slippage is more than expected by the user.
+
+## Allowed Tokens
+
+While the Dozer contract may allow many tokens for swaps we will only have a pre-approved set of tokens shown on the wallet.
+
+The list of allowed tokens will be served via CDN to the wallet, the document will contain the list of tokens allowed in each network, each token will also have a link to the icon (for the UI) that will also be available via CDN.
+
+Failure to resolve the list of tokens will block the swap screen from fully loading.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Token Swap UI
+
+The token swap button on the main screen will be loaded depending on a feature flag, once clicked the allowed token list will be fetched during an initial load step before the user can interact with the swap form.
+
+The screen should track if the user last changed the input token or the output token, this information is important to determine which method to call from the contract and how the swap will be made.
+
+### Choosing tokens for the swap
+
+Once the user selects both tokens and writes a value on either input or output token area the wallet will start a cancellable async task to fetch the swap information from the contract.
+
+The fetching of the swap information has to be a cancellable async task because the user may change the tokens during the request so we will need to stop the current request and start a second one with the new tokens chosen.
+
+When the exchange rate is known we will be able to calculate the desired values for the swap.
+These values will be locked in when the user clicks the review button.
+
+## Dozer token swap protocol methods
+
+### User defined the input tokens
+
+We will call the following method to receive the swap information.
+
+`find_best_swap_path(self, amount_in: Amount, token_in: TokenUid, token_out: TokenUid, max_hops: int)`
+- `amount_in`: The amount of input tokens
+- `token_in`: The input token uid
+- `token_out`: The output token uid
+- `max_hops`: Maximum number of hops (1-3, but algorithm handles any number)
+
+The path will define the next method to call.
+if only 1 swap is detected on the path `swap_exact_tokens_for_tokens(self, ctx: Context, fee: Amount)` should be called, the input and output token and amounts should be defined in the actions.

```

Comment:
question: It's not clear to me what happens in this case.

Will the user call for `find_best_swap_path()`, through the screen, actually end up internally executing a transaction that will swap the tokens using `swap_exact_tokens_for_tokens()` without further input?

If so, will this have a chance to end up creating outstanding balances for the user? How will the mobile wallet interface deal with the possibility of leaving unredeemed tokens inside the contract?